sensing/src/main/java/com/google/android/sensing/Authenticator.kt

```diff
@@ -17,39 +17,73 @@
 package com.google.android.sensing
 
 import androidx.annotation.WorkerThread
+import com.google.android.sensing.upload.AssumeRoleWithCustomTokenBasedCredentialProvider
+import io.minio.credentials.Credentials
 import io.minio.credentials.Provider
+import okhttp3.OkHttpClient
 
 @WorkerThread
 /**
- * [SensingEngine] depends on the developer app to handle user's authentication. The developer
- * application may provide the implementation during the [SensingEngine] initial setup.
+ * Sensing library depends on the developer app to handle user's authentication. The developer
+ * application may provide the implementation during the initial setup.
  *
- * We provide 3 interfaces:
- * 1. BasicAuthenticator: Accepts access-key (username) and secret-key (password) directly
- * 2. CredentialProviderAuthenticator: Accepts io.minio.credentials.Provider which already has
- * implementations like io.minio.credentials.LdapIdentityProvider
- * 3. CustomIDPAuthenticator: Accepts token and other details required for minio server to generate
- * temporary credentials, ie access and secret keys. Supports a custom external identity manager
- * using the MinIO Authentication Plugin extension.
+ * Application may use one of the abstract implementations of this interface or provide its own
+ * implementation. The interface returns a [io.minio.credentials.Provider] which is used to build a
+ * [MinioAsyncClient].
+ * 1. [MinioInternalIDPAuthenticator]: Implement this if deployment supports built-in IDP. Accepts
+ * access-key (username) and secret-key (password) directly.
+ * 2. [MinioIDPPluginAuthenticator]: Implement this if deployment supports external IDP. Accepts
+ * [OkHttpClient], [MinioIDPPluginAuthenticator.TokenProvider], roleArn, durationSeconds and
+ * stsEndpoint to create a [AssumeRoleWithCustomTokenBasedCredentialProvider].
  */
-interface Authenticator
+interface Authenticator {
+  fun getCredentialsProvider(): Provider
+}
 
-interface BasicAuthenticator : Authenticator {
+abstract class MinioInternalIDPAuthenticator : Authenticator {
```

Comment:
nit: Is `MinioExternalIDPAuthenticator` a good name?

Reply:
I guess you mean this change for `MinioIDPPluginAuthenticator`.

External there are other options like ldap and openid.
This particular implementation supports external custom idps which are supported via minio idp plugin.